source/common/protobuf/utility.cc

```diff
@@ -596,6 +597,156 @@ std::string MessageUtil::CodeEnumToString(ProtobufUtil::error::Code code) {
   }
 }
 
+namespace {
+
+// Forward declaration for mutually-recursive helper functions.
+void redact(Protobuf::Message* message, bool ancestor_is_sensitive);
+
+using Transform = std::function<void(Protobuf::Message*, const Protobuf::Reflection*,
+                                     const Protobuf::FieldDescriptor*)>;
+
+// To redact opaque types, namely `Any` and `TypedStruct`, we have to reify them to the concrete
+// message types specified by their `type_url` before we can redact their contents. This is mostly
+// identical between `Any` and `TypedStruct`, the only difference being how they are packed and
+// unpacked. Note that we have to use reflection on the opaque type here, rather than downcasting
+// to `Any` or `TypedStruct`, because any message we might be handling could have originated from
+// a `DynamicMessageFactory`.
+bool redactOpaque(Protobuf::Message* message, bool ancestor_is_sensitive,
+                  absl::string_view opaque_type_name, Transform unpack, Transform repack) {
+  // Ensure this message has the opaque type we're expecting.
+  if (message->GetDescriptor()->full_name() != opaque_type_name) {
+    return false;
+  }
+
+  // Enumerate the fields of this opaque type. We expect to find exactly two: `type_url` and
+  // `value`, but not necessarily in that order.
```

Comment:
Why can't we use `FindFieldByName`? I think that would be a bit clearer.

Reply:
Done. Thanks for that -- I hadn't discovered that chunk of the API -- it's a huge improvement.